### DISCUSSION TITLE
Thoughts on avoid duplicates?

### DISCUSSION THREAD

**Discussion by joshcanhelp** [SitanHuang/ledg#8] [Ideas]
I noticed that, during testing, I can import the same records multiple times. All of my transactions will be imported and some of the accounts I use include a transaction ID in the export. If the same ID comes in again, it should warn and allow me to continue if I want. Otherwise, transactions with the same date, amount, and description could warn as well.

If that doesn't go against any fundamentals of this project and it doesn't require a huge refactor, I could take a crack at a PR to add it.

**Comment by SitanHuang** [comment]
It is a very useful feature indeed. ledg itself has those 8 character random IDs, but since transaction ID and other attributes are user-defined, I think this responsibility of checking duplicates lies in the parser - the user code. What ledg needs to do is perhaps to provide macros or at least a way to execute the asynchronous [query_exec()](https://github.com/SitanHuang/ledg/blob/develop/lib/core/query.js#L22) inside parsers, which means the `ctx.process()` macro must accept both sync(for backwards compatibility) and async callbacks. 

One way is to let `ctx.process` and `fs_read_csv` accept async user callbacks that check for duplicates via `query_exec`; but `cmd_add`(where your data is actually added) is executed after `import_csv_exec` so duplicates inside each run aren't detected. Another way is to execute `cmd_add` inside `import_csv_exec` but that means `bail=true` won't discard all previous imports(right now, I believe if parser returns error for one of the rows, and bail is set to true, the whole import transaction is stopped and nothing is added).

Both methods aren't against fundamentals of this project and do not require huge refactors. A PR of either implementation would be great! But do let me know which one you prefer, or if you have other implementation ideas to have both `bail=true` stop whole import transaction AND check for duplicates inside each run.

Thank you.

**Reply by joshcanhelp** [reply]
Thank you! I will take a closer look. I realized that I needed to categorize everything before importing so I'm working on that (boring) task now!

**Comment by SitanHuang** [comment]
@joshcanhelp a critical error had been discovered as a result of the ending of daylight savings time. I had pushed a commit and please AVOID using old versions of ledg to prevent dates being corrupted!
